daceml/autodiff/pytorch.py

```diff
@@ -0,0 +1,200 @@
+
+        :param module: the parent pytorch module.
+        :param model: the model to convert.
+        :param apply_strict: whether to apply strict transformations before creating the backward pass.
+        :return: the PyTorch compatible :class:`torch.autograd.Function`.
+    """
+
+    if len(model.sdfg.nodes()) != 1:
+        raise AutoDiffException(
+            "Expected to find exactly one SDFGState, found {}".format(
+                len(model.sdfg.nodes())))
+
+    forward_sdfg = model.sdfg
+    forward_state = model.sdfg.nodes()[0]
+
+    backward_sdfg = dace.SDFG(forward_sdfg.name + "_backward")
+    backward_state = backward_sdfg.add_state()
+
+    gen = BackwardPassGenerator(
+        sdfg=forward_sdfg,
+        state=forward_state,
+        given_gradients=[clean_onnx_name(name) for name in model.outputs],
+        # TODO filter inputs that don't require grad somehow...
+        required_gradients=[clean_onnx_name(name) for name in model.inputs],
+        backward_sdfg=backward_sdfg,
+        backward_state=backward_state,
+        apply_strict=apply_strict)
+
+    backward_result, backward_grad_arrays, backward_input_arrays = gen.backward(
+    )
+
+    replaced_scalars = {}
+    for name, desc in backward_input_arrays.items():
+        if name not in forward_sdfg.arrays:
+            raise AutoDiffException(
+                "Expected to find array with name '{}' in SDFG".format(name))
+
+        forward_desc = forward_sdfg.arrays[name]
+        # we will save this output and pass it to the backward pass
+
+        # Views should not be forwarded. Instead the backward pass generator should forward the source of the view,
+        # and rebuild the sequence of required views in the backward pass.
+        assert type(forward_desc) is not dt.View
+        if type(forward_desc) is dt.Scalar:
+            # we can't return scalars from SDFGs, so we add a copy to an array of size 1
+            arr_name, _ = forward_sdfg.add_array(name + "_array", [1],
+                                                 forward_desc.dtype,
+                                                 transient=False,
+                                                 find_new_name=True)
+            copy_state = forward_sdfg.add_state_after(forward_state,
+                                                      label="copy_out_" +
+                                                      arr_name)
+            copy_state.add_edge(copy_state.add_read(name), None,
+                                copy_state.add_write(arr_name), None,
+                                dace.Memlet(name + "[0]"))
+            replaced_scalars[name] = arr_name
+        else:
+            forward_sdfg.arrays[name].transient = False
+
+    backward_sdfg.validate()
+
+    class DaceFunction(torch.autograd.Function):
+        _backward_sdfg = backward_sdfg
+        _forward_model = model
+        _backward_result = backward_result
+
+        @staticmethod
+        def forward(ctx, *inputs):
```

Comment:
I'm assuming this will be bad performance-wise. the forward function is on the critical path and you'd want to move everything you can out of it (to the point where I want to remember the order of parameters and fast-call SDFGs even). We'll see how it performs, but consider moving things out of forward and backward

Reply:
Yup defintely not good for performance, but this code is subject to change anyway since we still need to figure out how we can control allocation of parameters